### ISSUE TITLE
Issue with keras Batch Normalization

### ISSUE THREAD

**Issue by tomoleary** [tomoleary/hessianlearn#5]
Need to be able to handle reshaping trainable variables when bn is used. This is done in the constructor for problem.

**Comment by tomoleary** [comment]
Appears not to be an issue with BN, instead it is an issue with the automatic differentiation of the resnet models.

**Comment by tomoleary** [comment]
There is an issue with differentiating BatchNormalization when used right after a convolutional layer, but no issue with dense evidently. This issue arises when attempting to take derivatives with respect to `tf.compat.v1.trainable_variables`, which is deprecated in tf2
This exposes the issue:

```
classifier = tf.keras.Sequential([
	tf.keras.layers.Conv2D(16, kernel_size=3, activation='relu', input_shape=(28,28,1)),
	tf.keras.layers.BatchNormalization(),
	tf.keras.layers.Flatten(),
	tf.keras.layers.Dense(10, activation='softmax')
	])
```

**Comment by tomoleary** [comment]
This is a potentially viable workaround for batch normalization with convolution:

```
classifier = tf.keras.Sequential()
classifier.add(tf.keras.layers.Conv2D(16, kernel_size=3, activation='relu', input_shape=(28,28,1)))
conv_out_shape = classifier.output_shape[1:]
classifier.add(tf.keras.layers.Flatten())
classifier.add(tf.keras.layers.BatchNormalization())
classifier.add(tf.keras.layers.Reshape(conv_out_shape))
classifier.add(tf.keras.layers.Flatten())
classifier.add(tf.keras.layers.Dense(10, activation='softmax'))
```
😧

**Comment by tomoleary** [comment]
This is a problem when `tf.keras.model.Model.trainable_weights` is used as well.
